logprep/ng/sender.py

```diff
@@ -26,44 +26,47 @@ def __str__(self) -> str:
 class Sender(Iterator):
     """Sender class to handle sending events to configured outputs."""
 
-    def __init__(self, pipeline: Pipeline, outputs: list[Output], error_output: Output) -> None:
-        self._events = (event for event in pipeline if event is not None and event.data)
+    def __init__(
+        self,
+        pipeline: Pipeline,
+        outputs: list[Output],
+        error_output: Output,
+        process_count: int = 3,
+    ) -> None:
+        self.pipeline = pipeline
         self._outputs = {output.name: output for output in outputs}
         self._default_output = [output for output in outputs if output.default][0]
         self._error_output = error_output
-        self.batch_size = getattr(self._default_output._config, "message_backlog_size", 10)
+        self.batch_size = process_count
         self.batch: list[LogEvent] = []
+        self.should_exit = False
 
     def __next__(self) -> LogEvent | ErrorEvent:
-        event = next(self._events)
-        if event.state == EventStateType.PROCESSED:
-            self._send_processed(event)
-            for output in self._outputs.values():
-                output.flush()
-        if self._error_output and event.state == EventStateType.FAILED:
-            event = self._send_failed(event)
-            self._error_output.flush()
-        if isinstance(event, ErrorEvent) and event.state == EventStateType.FAILED:
-            logger.error("Error during sending to error output: %s", event)
-        return event
+        """not implemented, use iter()"""
+        raise NotImplementedError("Use iter() to get events from the Sender.")
 
     def __iter__(self) -> Generator[LogEvent | ErrorEvent, None, None]:
         """Iterate over processed events."""
-        while True:
+        while 1:
+            logger.debug("Sender iterating")
             self.batch.clear()
-            self.batch += list(islice(self._events, self.batch_size))
-            if not self.batch:
-                break
+            self.batch += list(islice(self.pipeline, self.batch_size))
             self._send_and_flush_processed_events()
             if self._error_output:
                 self._send_and_flush_failed_events()
-
+            if self.should_exit:
+                logger.debug("Sender exiting")
+                return
```

Comment:
Isn’t that actually what we want here?

```suggestion
                raise StopIteration
```